flink-runtime/src/main/java/org/apache/flink/streaming/api/graph/AdaptiveGraphManager.java

```diff
@@ -0,0 +1,697 @@
+            }
+        }
+        return getChainEntryPoints();
+    }
+
+    private Map<Integer, OperatorChainInfo> getChainEntryPoints() {
+
+        final Map<Integer, OperatorChainInfo> chainEntryPoints = new HashMap<>();
+        Iterator<Map.Entry<Integer, OperatorChainInfo>> iterator =
+                pendingChainEntryPoints.entrySet().iterator();
+
+        while (iterator.hasNext()) {
+            Map.Entry<Integer, OperatorChainInfo> entry = iterator.next();
+            Integer startNodeId = entry.getKey();
+            OperatorChainInfo chainInfo = entry.getValue();
+            if (!isReadyToCreateJobVertex(chainInfo)) {
+                continue;
+            }
+            chainEntryPoints.put(startNodeId, chainInfo);
+            iterator.remove();
+        }
+        return chainEntryPoints;
+    }
+
+    /**
+     * Responds to calculating the forward group and reset the parallelism of all nodes in the same
+     * forward group to make them the same.
+     *
+     * @param sourceNodes the source nodes.
+     */
+    private void setVertexParallelismsForDynamicGraphIfNecessary(List<StreamNode> sourceNodes) {
+        Map<Integer, List<StreamEdge>> transitiveNonChainableOutEdgesMap = new HashMap<>();
+        Map<StreamNode, List<StreamNode>> topologicallySortedChainedStreamNodesMap =
+                new LinkedHashMap<>();
+        Set<Integer> visitedStartNodes = new HashSet<>();
+
+        List<StreamNode> enterPoints =
+                getEnterPoints(sourceNodes, topologicallySortedChainedStreamNodesMap);
+
+        for (StreamNode streamNode : enterPoints) {
+            traverseFullGraph(
+                    streamNode.getId(),
+                    streamNode.getId(),
+                    transitiveNonChainableOutEdgesMap,
+                    topologicallySortedChainedStreamNodesMap,
+                    visitedStartNodes);
+        }
+        computeForwardGroupAndSetNodeParallelisms(
+                transitiveNonChainableOutEdgesMap, topologicallySortedChainedStreamNodesMap);
+    }
+
+    private List<StreamNode> getEnterPoints(
+            List<StreamNode> sourceNodes, Map<StreamNode, List<StreamNode>> chainedStreamNodesMap) {
+        List<StreamNode> enterPoints = new ArrayList<>();
+        for (StreamNode sourceNode : sourceNodes) {
+            if (isSourceChainable(sourceNode, streamGraph)) {
+                StreamEdge outEdge = sourceNode.getOutEdges().get(0);
+                StreamNode startNode = streamGraph.getStreamNode(outEdge.getTargetId());
+                chainedStreamNodesMap
+                        .computeIfAbsent(startNode, k -> new ArrayList<>())
+                        .add(sourceNode);
+                enterPoints.add(startNode);
+            } else {
+                chainedStreamNodesMap.computeIfAbsent(sourceNode, k -> new ArrayList<>());
+                enterPoints.add(sourceNode);
+            }
+        }
+        return enterPoints;
```

Comment:
How about to let this method simply return `chainedStreamNodesMap`? In this way we do not need to create and maintain the `enterPoints`. And the `chainedStreamNodesMap` can be created within this method instead of being a parameter. Also , maybe the name can be `createChainEntryPoints()`?